table/block_based/block_based_table_reader.cc

```diff
@@ -987,20 +987,56 @@ Status BlockBasedTable::PrefetchIndexAndFilterBlocks(
 
   const bool use_cache = table_options.cache_index_and_filter_blocks;
 
-  // pin both index and filters, down to all partitions.
-  const bool pin_all =
-      rep_->table_options.pin_l0_filter_and_index_blocks_in_cache &&
+  const bool maybe_flushed =
       level == 0 && file_size <= max_file_size_for_l0_meta_pin;
+  std::function<bool(PinningTier, PinningTier)> is_pinned =
+      [maybe_flushed, &is_pinned](PinningTier pinning_tier,
+                                  PinningTier fallback_pinning_tier) {
+        // Fallback to fallback would lead to infinite recursion. Disallow it.
+        assert(fallback_pinning_tier != PinningTier::kFallback);
+
+        switch (pinning_tier) {
+          case PinningTier::kFallback:
+            return is_pinned(fallback_pinning_tier,
+                             PinningTier::kNone /* fallback_pinning_tier */);
+          case PinningTier::kNone:
+            return false;
+          case PinningTier::kMaybeFlushed:
+            return maybe_flushed;
+          case PinningTier::kAll:
+            return true;
+        };
+
+        // In GCC, this is needed to suppress `control reaches end of non-void
+        // function [-Werror=return-type]`.
+        assert(false);
+        return false;
+      };
+  const bool pin_metablock_top_level_index = is_pinned(
+      table_options.cache_pinning_options.metablock_top_level_index_pinning,
+      table_options.pin_top_level_index_and_filter ? PinningTier::kAll
+                                                   : PinningTier::kNone);
+  const bool pin_metablock_partition =
+      is_pinned(table_options.cache_pinning_options.metablock_partition_pinning,
+                table_options.pin_l0_filter_and_index_blocks_in_cache
+                    ? PinningTier::kMaybeFlushed
+                    : PinningTier::kNone);
+  const bool pin_unpartitioned_metablock = is_pinned(
+      table_options.cache_pinning_options.unpartitioned_metablock_pinning,
+      table_options.pin_l0_filter_and_index_blocks_in_cache
+          ? PinningTier::kMaybeFlushed
+          : PinningTier::kNone);
```

Comment:
I think you can do this simpler by using the InitializeOptions method of BlockBasedTableFactory.  You should be able to update the cache_pinning_options in that method based on either of the settings.

Reply:
Additionally, if you update the pinning options in the InitializeOptions, you can mark the old/deprecated options as "kDontSerialize" and they will no longer be written to the output.  This would help make them more "obsolete".

Reply:
I will try this out but need to defer it until another PR since today's the last day for features in the next release, and there are still bugs to fix.